### ISSUE TITLE
aws.ebs - proposal of "stop-instance" action

### ISSUE THREAD

**Issue by llange** [cloud-custodian/cloud-custodian#3658]
In our usage (mainly tagging policy compliance actions) we have a need to stop EC2 instances attached to EBS volumes (the filters are on the EBS volumes).

Our use-case is something like (our real filter is more complex for the Tag Policy compliance part, but you get the idea. Also, there is notification, etc...):
```yaml
policies:
- name: ebs-tag-compliance-stop-instance
  resource: ebs
  filters:
    # => EBS older than 20 minutes
    - type: value
      key: CreateTime
      op: ge
      value_type: age
      value:
        minutes: 20
    # => EBS attached to running instances
    - type: instance
      key: State.Name
      op: in
      value: 
        - running
        - pending
    # => EBS attached to non-spot instances
    - or:
      - type: instance
        key: InstanceLifecycle
        value: absent
      - type: instance
        key: InstanceLifecycle
        op: not-equal
        value: "spot"
    # => EBS not properly tagged
    - or:
      - "tag:TagCompliance": absent
  actions:
    - stop-instance
```

To this end we've created the "stop-instance" EBS (volume) action that is shared in the attached Pull Request, for comments and feedback.

**Comment by kapilt** [comment]
Why not do this as an ec2 policy, instead of duplicating the stop action. The ec2 resource has an ebs filter

**Comment by kapilt** [comment]
Ie most of these filters are about ec2, this seems like it’s really an ec2 policy that for reasons that aren’t clear got written as an ebs policy

**Comment by llange** [comment]
You're right, and at first I did not know of the "ebs" filter for ec2 resources.

I absolutely can use this ebs filter for the same need and result ; it makes more sense.

The only (minor) issues that I have doing things "the right way" are:
- In terms of reporting (custodian report / c7n-org report), as I'm dealing with EC2 resources, I need to process all EBS subresources in order to find which one was not compliant (an EC2 may, and will, have multiple EBS volumes). The CSV output will not be so smooth to create (not impossible)
- In terms of policies, we are generating excerpts that we include in the main policies. Indeed, the "tag compliance policy" is a (long and complex) set of filters that we can include (using #2184) in each of the policies (to keep things DRY). This set of filters is mainly written with (lot of) "type: value" filters. When I'm dealing with EC2, this (included) excerpt applies to the EC2, not the attached EBS. To have it apply to the attached EBS, we need to duplicate this excerpt and rewrite it using "type: ebs" filter. It's less DRY.

Nevertheless, we will adapt to the "ebs" filter and get rid of this (proposed) "stop-instance" action, and find a way to generate the policies from templates (e.g. Jinja2) to keep things DRY and still have our tag compliance policy defined in an unique place.

(I still need to find a way to properly format the report, and I'm open to any suggestion)

Long story short : we can close both the issue and the PR ; either you do it or I can do it.

**Comment by kapilt** [comment]
wrt to cli csv report, --field="key=expr"  it is possible to do some summation in jmespath expressions, ie. sum volume sizes, or number of volumes per instance. It doesn't look like we currently annotate the ec2 instance with the volumes that matched on an ebs filter, so that may be needed as enhancement to enable that reporting.

if your already generating policies, then DRY is a feature of your generation not the output. even on inclusion it seems like its mostly about uniquely naming the common fragment. fwiw, on c7n-org there is also user defined variable support on a per account basis https://github.com/cloud-custodian/cloud-custodian/tree/master/tools/c7n_org#defining-and-using-variables

**Comment by llange** [comment]
I'll check what is possible to do with the CSV output. I was already using it "the other way" and the "ec2" filter of "ebs" provider was indeed anotating the ebs volumes with the instances that matched, so I was able to display the result in the CSV using "--field".
The other way around would still be possible ; I'll just have to confirm that:
- The "ebs" filter of "ec2" provider does indeed anotate (or else I'll enhance it)
- If only the matching volumes are annotated (or all the volumes) - I'd prefer to have only the matching volumes.
- What I can do to have a "beautiful" CSV, meaning that if more than 1 volume matched (and are annotated), I'll have to be creative in how I'll report the (mostly individual, not summed) info -using JMESPath- in the CSV column.

By the way, I have the same need in notification templates (mails, slack, ...) Is there the same JMESPath functionality available in the Jinja2 templating solution used there ? Or should we use the Python/Jinja2 vocabulary to express the same display needs as in the CSV with "--field" ?

For the policies, there were not already generated, but now I'm in the process of generating them ; so the issue is under control.

Thanks for the provided guidance.

**Comment by kapilt** [comment]
generally filters should annotate matched resources, a few do both, but the general balance is against enlarging the resource too much vs its base size, as we have size limits on notify transports (256kb), which we currently statically chunk the matched resources to.

re templates, thats interesting, it doesn't look like we expose jmespath there, but that sounds useful, worth a separate issue. note jinja has some builtin stuff for navigating nested structures, but the jmespath addition would be useful.

**Comment by kapilt** [comment]
filed a separate issue for the jmespath in templates support and closing this.

**Comment by llange** [comment]
Re filters and matches, I just submitted #3753 to enhance "EBS" filter of "EC2" ressource to have annotations of matched volumes.
